tap_mailchimp/streams.py

```diff
@@ -0,0 +1,545 @@
+import json
+import os
+import time
+import random
+import tarfile
+
+import singer
+from singer import metrics, metadata, Transformer
+from singer.utils import strptime_to_utc, should_sync_field
+from requests.exceptions import HTTPError
+
+LOGGER = singer.get_logger()
+
+MIN_RETRY_INTERVAL = 2 # 2 seconds
+MAX_RETRY_INTERVAL = 300 # 5 minutes
+MAX_RETRY_ELAPSED_TIME = 43200 # 12 hours
+
+# Break up reports_email_activity batches to iterate over chunks
+EMAIL_ACTIVITY_BATCH_SIZE = 100
+
+class BatchExpiredError(Exception):
+    pass
+
+def next_sleep_interval(previous_sleep_interval):
+    min_interval = previous_sleep_interval or MIN_RETRY_INTERVAL
+    max_interval = previous_sleep_interval * 2 or MIN_RETRY_INTERVAL
+    return min(MAX_RETRY_INTERVAL, random.randint(min_interval, max_interval))
+
+def get_abs_path(path):
+    return os.path.join(os.path.dirname(os.path.realpath(__file__)), path)
+
+def chunk_campaigns(sorted_campaigns, chunk_bookmark):
+    chunk_start = chunk_bookmark * EMAIL_ACTIVITY_BATCH_SIZE
+    chunk_end = chunk_start + EMAIL_ACTIVITY_BATCH_SIZE
+
+    if chunk_bookmark > 0:
+        LOGGER.info("reports_email_activity - Resuming requests starting at campaign_id %s (index %s) in chunks of %s",
+                    sorted_campaigns[chunk_start],
+                    chunk_start,
+                    EMAIL_ACTIVITY_BATCH_SIZE)
+
+    done = False
+    while not done:
+        current_chunk = sorted_campaigns[chunk_start:chunk_end]
+        done = len(current_chunk) == 0
+        if not done:
+            end_index = min(chunk_end, len(sorted_campaigns))
+            LOGGER.info("reports_email_activity - Will request for campaign_ids from %s to %s (index %s to %s)",
+                        sorted_campaigns[chunk_start],
+                        sorted_campaigns[end_index - 1],
+                        chunk_start,
+                        end_index - 1)
+            yield current_chunk
+        chunk_start = chunk_end
+        chunk_end += EMAIL_ACTIVITY_BATCH_SIZE
+
+def transform_activities(records):
```

Comment:
If this function is used for any one specific stream then keep it with that class only.

Reply:
As this function is not using any class' variable, the pylint is failing with `no-self-use`. Thus, separated this function.